stablehlo/reference/Index.cpp

```diff
@@ -20,17 +20,56 @@ limitations under the License.
 namespace mlir {
 namespace stablehlo {
 
-LogicalResult verifyIndex(llvm::ArrayRef<int64_t> shape,
-                          llvm::ArrayRef<int64_t> index) {
+Index Index::operator+(const Index &other) const {
+  if (size() != other.size())
+    llvm::report_fatal_error("Index add expects operands of same size.");
+
+  SmallVector<int64_t> result;
+  for (auto [lhsIdx, rhsIdx] : llvm::zip(index_, other.index_))
+    result.push_back(lhsIdx + rhsIdx);
+  return Index(result);
+}
+
+Index Index::operator+(ArrayRef<int64_t> array) const {
+  if (size() != array.size())
+    llvm::report_fatal_error("Index add expects operands of same size.");
+
+  SmallVector<int64_t> result;
+  for (auto [lhsIdx, rhsIdx] : llvm::zip(index_, array))
+    result.push_back(lhsIdx + rhsIdx);
+  return Index(result);
```

Comment:
This can be deduplicated to `return *this + Index(array)`. Same below for `operator*`.